AK/HashTable.h

```diff
@@ -531,194 +461,169 @@ class HashTable {
         MUST(try_rehash(new_capacity));
     }
 
-    void rehash_in_place()
+    template<typename TUnaryPredicate>
+    [[nodiscard]] BucketType* lookup_with_hash(unsigned hash, TUnaryPredicate predicate) const
     {
-        // FIXME: This implementation takes two loops over the entire bucket array, but avoids re-allocation.
-        //        Please benchmark your new implementation before you replace this.
-        //        The reason is that because of collisions, we use the special "rehashed" bucket state to mark already-rehashed used buckets.
-        //        Because we of course want to write into old used buckets, but already rehashed data shall not be touched.
-
-        // FIXME: Find a way to reduce the cognitive complexity of this function.
-
-        for (size_t i = 0; i < m_capacity; ++i) {
-            auto& bucket = m_buckets[i];
-
-            // FIXME: Bail out when we have handled every filled bucket.
-
-            if (bucket.state == BucketState::Rehashed || bucket.state == BucketState::End || bucket.state == BucketState::Free)
-                continue;
-            if (bucket.state == BucketState::Deleted) {
-                bucket.state = BucketState::Free;
-                continue;
-            }
-
-            auto const new_hash = TraitsForT::hash(*bucket.slot());
-            if (new_hash % m_capacity == i) {
-                bucket.state = BucketState::Rehashed;
-                continue;
-            }
-
-            auto target_hash = new_hash;
-            auto const to_move_hash = i;
-            BucketType* target_bucket = &m_buckets[target_hash % m_capacity];
-            BucketType* bucket_to_move = &m_buckets[i];
+        if (is_empty())
+            return nullptr;
 
-            // Try to move the bucket to move into its correct spot.
-            // During the procedure, we might re-hash or actually change the bucket to move.
-            while (!is_free_bucket(bucket_to_move->state)) {
+        for (;;) {
+            auto* bucket = &m_buckets[hash % m_capacity];
+            if (bucket->state == BucketState::Free)
+                return nullptr;
```

Comment:
Potential optimization to try out: we could stop the probe when the bucket's probe length stops incrementing (i.e. is less than the expected probe length).

Reply:
Adding a check for the probe length somehow made things slower, so I'm leaving this as-is. CC @Hendiadyoin1

Reply:
What I mean is, if we keep track of the current probe-length, we can compare it to the bucket's probe-length, for a potentially cheaper comparison
This will likely be more relevant if we have a separate state-storage as proposed on discord

Can you try that (without the separate storage) and see if that makes a difference?

Reply:
I did exactly that (not with separate storage), and it resulted in slightly degraded performance...

Reply:
Well then, might be a thing that depends on the load-factor and comparison cost
Maybe leave this as a note for later investigation, or leave it be

(the first statement sounded like a different check)